### ISSUE TITLE
monome ops need added focus functionality

### ISSUE THREAD

**Issue by tehn** [tehn/aleph#169]
- [x] do not send LED data when no focus
- [x] clear and update LEDs on focus

**Comment by ghost** [comment]
So having thought about this a bit I think the best way to handle the focus-switching is for 'focus' inputs on these monolithic grid ops to be stateless.  By that I mean they always display 0 on BEES inputs page.
- Whenever the focus input for an op is banged with _any_ number, that op grabs focus from whoever currently is holding focus.
- 'having focus' means that op will receive button presses & it has 'write access' to the grid LEDs.
- for starters, every op can transmit LED data regardless of focus, & the driver can ignore all but the focused op.
- each grid op must store the state of all it's _internal_ grid LEDs independent of actual values on the physical grid
- when an op grabs focus, it updates the whole grid to match up with it's internal LED state
- ops do not have to be 'notified' when they lose focus

A possible optimisation (which adds some complexity) would be to also have the backend driver notify an op when it loses focus.  That way it can simply _not send_ the LED data to backend driver...  But I have a feeling this may not make a _huge_ difference to performance.

Pretty sure that plan will work, and work well, though carrying out the plan probably more challenging than maybe I'd like!  @tehn, @catfact either of you see any pitfalls to this method of swapping grid focus between apps?

**Comment by catfact** [comment]
> Whenever the focus input for an op is banged with any number, that op grabs focus from whoever currently is holding focus.

that seems ok. it has drawbacks: 
- mainly, it's a mite confusing for UI. you can't scroll through the inputs list and see who has focus.
- it seems relatively easy to accidentally bang something, from preset recall or on scene load, that you didn't mean to. when the input value is 1/0, accidental bangs don't matter as long as the _value_ is correct.

> each grid op must store the state of all it's internal grid LEDs independent of actual values on the physical grid

agreed, that's necessary, one way or another, for focus switching to work well. it can't work now. 

> it updates the whole grid to match up with it's internal LED state

this i'm a bit unsure on. i'm coming around to what i think @scanner-darkly was proposing on the forum, which is that the driver needn't keep its own state at all (except as a fallback); instead it points at the buffer of whatever op has focus. 

otherwise, seems like the options are:
1. periodically copy from the focussed op, from the refresh timer i guess. as you say this shouldn't really be a big deal, but it's not a lovely solution.
2. force the operators to explicitly request refreshes. do the copy at that point, set dirty-flag for appropriate frames, and wait for the refresh timer for bits to actually move out. drawback: operator might ask for way more copy steps than is really necessary.

but i'm still not really seeing the benefit of copying since the driver presumably needs to know about the focused op address anyway.

**Comment by ghost** [comment]
> that seems ok. it has drawbacks: mainly, it's a mite confusing for UI. you can't scroll through the inputs list and see who has focus.
>
Yes I thought harder about this and come round to your point of view on this.  Additionally:
https://github.com/rick-monster/aleph/blob/dev/apps/bees/src/net_monome.c#L65
This focus flag can be replaced by a pointer directly into the operator's focus input - so the glue layer will be responsible for updating the focus input when one op steals focus from another.
> this i'm a bit unsure on. i'm coming around to what i think @scanner-darkly was proposing on the forum, which is that the driver needn't keep its own state at all (except as a fallback); instead it points at the buffer of whatever op has focus.
> 
Sounds like the right thing to me - gonna make the necessary changes to driver on dev branch...  (i.e putting the avr32lib merge to one side for a minute)

will pull-request this change back into avr32lib along with my serial comms changes on dev at a later stage.
